text/0000-return-position-impl-trait-in-traits.md

```diff
@@ -0,0 +1,508 @@
+[motivation]: #motivation
+
+The `impl Trait` syntax is currently accepted in a variety of places within the Rust language to mean "some type that implements `Trait`" (for an overview, see the [explainer] from the impl trait initiative). For function arguments, `impl Trait` is [equivalent to a generic parameter][apit] and it is accepted in all kinds of functions (free functions, inherent impls, traits, and trait impls).
+
+In return position, `impl Trait` [corresponds to an opaque type whose value is inferred][rpit]. In that role, it is currently accepted only in free functions and inherent impls. This RFC extends the support for return position `impl Trait` in functions in traits and trait impls.
+
+[explainer]: https://rust-lang.github.io/impl-trait-initiative/explainer.html
+[apit]: https://rust-lang.github.io/impl-trait-initiative/explainer/apit.html
+[rpit]: https://rust-lang.github.io/impl-trait-initiative/explainer/rpit.html
+
+## Example use case
+
+The use case for `-> impl Trait` in trait functions is similar to its use in other contexts: traits often wish to return "some type" without specifying the exact type. As a simple example that we will use through the RFC, consider the `NewIntoIterator` trait, which is a variant of the existing `IntoIterator` that uses `impl Iterator` as the return type:
+
+```rust
+trait NewIntoIterator {
+    type Item;
+    fn into_iter(self) -> impl Iterator<Item = Self::Item>;
+}
+```
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+*This section assumes familiarity with the [basic semantics of impl trait in return position][rpit].*
+
+When you use `impl Trait` as the return type for a function within a trait definition or trait impl, the intent is the same: impls that implement this trait return "some type that implements `Trait`", and users of the trait can only rely on that.
+
+<!--However, the desugaring to achieve that effect looks somewhat different than other cases of impl trait in return position. This is because we cannot desugar to a type alias in the surrounding module; we need to desugar to an associated type (effectively, a type alias in the trait).-->
+
+Consider the following trait:
+
+```rust
+trait IntoIntIterator {
+    fn into_int_iter(self) -> impl Iterator<Item = u32>;
+}
+```
+
+The semantics of this are analogous to introducing a new associated type within the surrounding trait;
+
+```rust
+trait IntoIntIterator { // desugared
+    type IntoIntIter: Iterator<Item = u32>;
+    fn into_int_iter(self) -> Self::IntoIntIter;
+}
+```
+
+When using `-> impl Trait`, however, there is no associated type that users can name.
+
+By default, the impl for a trait like `IntoIntIterator` must also use `impl Trait` in return position.
+
+```rust
+impl IntoIntIterator for Vec<u32> {
+    fn into_int_iter(self) -> impl Iterator<Item = u32> {
+        self.into_iter()
+    }
+}
+```
+
+It can, however, give a more specific type with `#[refine]`:[^refine]
```

Comment:
Is there a rationale given somewhere for requiring `#[refine]`?

Reply:
https://github.com/rust-lang/rfcs/pull/3245

Reply:
I think the RFC includes the rationale and is linked to by the footnote, so resolving. Feel free to unresolve if you think there should be more explicit text here.